libraries/ESP8266HTTPClient/src/ESP8266HTTPClient.cpp

```diff
@@ -1225,41 +1226,49 @@ bool HTTPClient::sendHeader(const char * type)
         return false;
     }
 
-    String header = String(type) + ' ' + (_uri.length() ? _uri : F("/")) + F(" HTTP/1.");
+    String header;
+    // 168: Arbitrarily chosen to have enough buffer space for avoiding internal reallocations
+    header.reserve(_headers.length() + _base64Authorization.length() + _host.length() + 168);
```

Comment:
Choosing some estimated upper bound like this is fine, but how about doing a finer estimate here? You have at hand at least the following:
- _uri.length(), which in theory could be up to 2048 chars max, although an appropriate length is considered to be 75
- _host.length() which could be up to 255
- userAgent.length() I have no idea what the max length is, but a quick search shows upwards of 100

Of course, you could then reduce your 168 to something smaller.

Reply:
There is a trade-off to make: correctness of the calculation vs code size overhead for the correct calculation. you're correct that we can add _uri.length and userAgent.length() that can be added, but lets not overdo it because it really doesn't matter whether we'll have one realloc() or zero realloc(). the important improvement here is that we no longer have on average 4-5 realloc()s.